ddclient.in

```diff
@@ -2020,114 +2022,176 @@ sub geturl {
-    $request .= $data;
-    $rq .= $data;
-
-    my %socket_args = (
-        PeerAddr => $peer,
-        PeerPort => $port,
-        Proto => 'tcp',
-        MultiHomed => 1,
-        Timeout => opt('timeout'),
-    );
-    my $socket_class = 'IO::Socket::INET';
-    if ($use_ssl) {
-        # IO::Socket::SSL will load IPv6 support if available on the system.
-        load_ssl_support;
-        $socket_class = 'IO::Socket::SSL';
-        $socket_args{SSL_ca_file} = opt('ssl_ca_file') if defined(opt('ssl_ca_file'));
-        $socket_args{SSL_ca_path} = opt('ssl_ca_dir') if defined(opt('ssl_ca_dir'));
-    } elsif ($globals{'ipv6'} || $ipversion eq '6') {
-        load_ipv6_support;
-        $socket_class = 'IO::Socket::INET6';
-    }
-    if ($ipversion eq '4') {
-        $socket_args{Domain} = PF_INET;
-        $socket_args{Family} = AF_INET;
-    } elsif ($ipversion eq '6') {
-        $socket_args{Domain} = PF_INET6;
-        $socket_args{Family} = AF_INET6;
-    } elsif ($ipversion ne '') {
-        fatal("geturl passed unsupported 'ipversion' value %s", $ipversion);
-    }
+    if (opt('curl')) {
+        if (!`curl --version 2>/dev/null`) {
+            fatal("You requested network access with curl but curl is not installed.");
+        }
+        ## Access network using curl.
+        if (!opt('exec')) {
+            debug("skipped network connection");
+            verbose("SENDING:", "%s", "${server}/${url}");
+        } elsif (($headers ne '')) {
+            my $timeout    = opt('timeout');
+            my $curlopt  = '';
+            my @head_lines = split('\n', $headers);
+            my $head_args  = '';
+            foreach my $head_line (@head_lines) {
+                $head_args .= "-H '$head_line' ";
+            }
+            $curlopt = '--ipv4' if ($ipversion eq '4');
+            $curlopt = '--ipv6' if ($ipversion eq '6');
+
+            # don't include ${url} as that might expose login credentials
+            $0 = sprintf("%s - curl (w/headers) sending to %s", $program, "https://${server}");
+            verbose("SENDING:", "curl (w/headers) to %s", "https://${server}");
+            $reply = <<`USE_CURL_WITH_HEADERS`;
+curl -s -X $method --user-agent '${program}/${version}' \\
+              --connect-timeout $timeout --max-time $timeout $curlopt \\
+              ${head_args} \\
+              --data '${data}' \\
+              --url 'https://${server}/${url}' 2>/dev/null
+USE_CURL_WITH_HEADERS
+
+            if (!$reply) {
+                # don't include ${url} as that might expose login credentials
+                warning("curl (w/headers) cannot connect to https://%s using IPv%s",${server},$ipversion);
+            }
+        } else {
+            my $timeout  = opt('timeout');
+            my $auth     = '';
+            my $curlopt  = '';
+            my $protocol = 'http';
+            $auth    = "--user '${login}:${password}'" if (($login ne '') && ($password ne ''));
```

Comment:
Don't pass secrets in command-line arguments—use an environment variable, stdin, or a file instead.

Reply:
Not sure how to accomplish this.  Maybe the `--netrc-file` option to curl?  Open to suggestions.

Reply:
I think it may be possible to use:
```
xargs curl -s  --url 'http://example.com/' <<< "--user '${login}:${password}'"
```
However this will require opening for both reading and writing... which means `use IPC::Open2`. I have not managed to get curl on its own with `-K-` option to work -- at least not yet.

Reply:
Reason behind this is probably that you can read the secrets from other programs like htop.

Reply:
> Reason behind this is probably that you can read the secrets from other programs like htop.

Yes exactly.  But piping in from STDIN should close that loophole.  I'm looking into how to get that working.

Reply:
Running
```
xargs curl ... <<<"--user username:password"
```
is just a complicated way of running
```
curl ... --user username:password
```
It still passes the secret via a command-line argument so it will still show up in `ps` and whatnot. Also, `<<<` is a bashism—it's not available in POSIX shell.

The `--config` (`-K`) command-line option is probably what you want, unless curl can take the password via an environment variable.

Reply:
Yes, it is now implemented using config file.  See latest code.